Add codemods: React.js

jsnmoon opened the pull request:
Part of #11688.

This change adds codemods in `bin/codemods` for porting legacy React.js code.
- `react-create-class`: Safely replaces `React.createClass` with `React.Component` or `createReactClass` from `'create-react-class'`. [(source)](https://github.com/reactjs/react-codemod/#explanation-of-the-new-es2015-class-transform-with-property-initializers)
- `react-prop-types`: Replaces `React.PropTypes` with `prop-types`; applicable for React 15.5+ where [React.PropTypes](https://facebook.github.io/react/blog/2017/04/07/react-v15.5.0.html#migrating-from-react.proptypes) has been extracted to a separate package. [(source)](https://github.com/reactjs/react-codemod/#react-proptypes-to-prop-types)

Please note that this currently uses by [personal fork](https://github.com/jsnmoon/react-codemod/tree/2017-06-12-fix) with couple of bug fixes on top of react-codemod. The following PRs have been filed upstream:

- reactjs/react-codemod#122
- reactjs/react-codemod#132
- reactjs/react-codemod#156
- reactjs/react-codemod#157

Todo:
- [x] Fix class codemod bug erroring out on trailing comma (see [this issue](https://github.com/reactjs/react-codemod/pull/122))
- [x] Thoroughly test class codemod on `client/`
- [x] Resolve issues with class codemod (Most pressing issue has been resolved
- [x] Thoroughly test prop-types codemod on `client/`

Issues:
- `react-create-class` codemod
  - [x] Incorrect removes `react` import for JSX files when falling back to `createClass` import
    - [x] ex. `client/post-editor/media-modal/gallery/preview-shortcode.jsx` (react import unnecessarily removed)
    - [x] ex. `client/post-editor/editor-ground-control/index.jsx` (multiple imports are removed)
    - [x] ex. `client/my-sites/themes/thanks-modal.jsx` (malformed import after transform)
  - Does not assign a class name if React.createClass was directly exported
    - Noted - not significant enough to block merge
    - ex. `client/post-editor/media-modal/gallery/remove-button.jsx`
  - Favors CommonJS import notation if the file contains any CommonJS imports
    - Noted - not significant enough to block merge
    - ex. `client/post-editor/post-editor.jsx`

ockham commented:
> - `react-create-class`: Safely replaces `React.createClass` with `React.Component` or `createReactClass` from `'create-react-class'`. [(source)](https://github.com/reactjs/react-codemod/#explanation-of-the-new-es2015-class-transform-with-property-initializers)

How well does this work? Last time I tried it on the Calypso root dir I think it only changed some 4 components or so. (I'm speculating it might be due to our use of `this.translate()`, but I'm not sure.)

jsnmoon commented:
@ockham: I was under the impression it worked pretty reliably from my initial testing. To verify, I ran the `create-class` codemod on `client/` today and here are the results:

- Successfully processed: 699 files
  - Many cases of `React.createClass` were converted into classes extended from `React.Component`.
  - Many cases of `React.createClass` were converted into `createClass` imported from `createReactClass` package. These usually occur due to use of mixins other than `PureRenderMixin` or use of deprecated React API calls like `getDOMNode`, `isMounted`, `replaceProps`, `replaceState`, or `setProps`.
- Errored out: 39 files
  - I'm individually investigating these errors, but it'll take me some time to figure out exactly why these are happening - will update in the future.

Here are some issues I discovered from the successfully processed files:
1) The codemod does not assign a class name if `React.createClass` was directly exported
    - `export default React.createClass( ... )` becomes `export default class extends React.Component { ... }`
    - ex. `client/post-editor/media-modal/gallery/remove-button.jsx`
2) The codemod seems to favor CommonJS import notation if the file contains any CommonJS imports.
    - ex. `client/post-editor/post-editor.jsx`
3) If a `react` import is done with declarator siblings in CommonJS notation, the codemod  sometimes erroneously removes the entire import declaration.
    - My hunch is that the codemod doesn't properly handle CommonJS imports with sibling declarators. This issue can be skirted by converting these to ES module imports and then running the `react-create-class` codemod.
    - ex. `client/post-editor/media-modal/gallery/preview-shortcode.jsx` (`react` import unnecessarily removed)
    - ex. `client/post-editor/editor-ground-control/index.jsx` (multiple imports are removed)
    - ex. `client/my-sites/themes/thanks-modal.jsx` (malformed import after transform)

ockham commented on bin/codemods/react-prop-types:
```diff
@@ -0,0 +1,41 @@
+#!/usr/bin/env node
```
This file seems to be missing the executable bit currently.

jsnmoon replied:
Fixed!

gwwar commented on package.json:
```diff
@@ -184,6 +184,7 @@
     "nock": "8.0.0",
     "nodemon": "1.4.1",
     "react-addons-test-utils": "15.4.0",
+    "react-codemod": "github:reactjs/react-codemod#d6563d9",
```
Is there not a published version of this? Otherwise, github links can be shorter: 
```reactjs/react-codemod#d6563d9```

https://docs.npmjs.com/files/package.json#github-urls

ockham replied:
There doesn't seem to be an `npm` AFAICS. FWIW, [I added the devDep without SHA stamp](https://github.com/Automattic/wp-calypso/blob/121c5b4c5c459f4e40b491213fa385f546ccae3b/package.json#L191) (thus tracking `master`) a while back.

jsnmoon replied:
@gwwar Unfortunately, react-codemod isn't published on `npm` nor are releases properly tagged on their git repo, so I'll go ahead and shorten this dependency to your suggestion!

jsnmoon commented:
Rebased off of master and changed the react-codemod dependency to point at my fork containing [bug](https://github.com/jsnmoon/react-codemod/commit/f84e49e17cc14d0aca03f0aa92a2e78d9e06ec5c) [fixes](https://github.com/jsnmoon/react-codemod/commit/509179b255efe87a3f49e1c8eee865e925ab9ab4) [relevant](https://github.com/jsnmoon/react-codemod/commit/650171db5f316d4a98df8b9d2897547aec314f37) to Calypso.

ockham commented:
Rebased, and re-shrunk. This has been lingering for a while, so I think it'd be good to get it in soonish.

> Rebased off of master and changed the react-codemod dependency to point at my personal [fork](https://github.com/jsnmoon/react-codemod/tree/2017-06-12-fix) containing [bug](https://github.com/jsnmoon/react-codemod/commit/f84e49e17cc14d0aca03f0aa92a2e78d9e06ec5c) [fixes](https://github.com/jsnmoon/react-codemod/commit/509179b255efe87a3f49e1c8eee865e925ab9ab4) [relevant](https://github.com/jsnmoon/react-codemod/commit/650171db5f316d4a98df8b9d2897547aec314f37) to Calypso. 

Is this stuff going to be/has been fixed upstream? If so, it'd be obviously nice to continue to use upstream. Anyway, it'd be cool if you could link to upstream PRs from here so we can keep track.

One more bit of feedback: Since this PR contains a codemod to use the new standalone [`prop-types` npm](https://www.npmjs.com/package/prop-types), I think we should add it to `package.json`.

ockham commented:
> One more bit of feedback: Since this PR contains a codemod to use the new standalone [`prop-types` npm](https://www.npmjs.com/package/prop-types), I think we should add it to `package.json`.

Oh, same for `create-react-class`.

jsnmoon commented:
> Is this stuff going to be/has been fixed upstream? If so, it'd be obviously nice to continue to use upstream. Anyway, it'd be cool if you could link to upstream PRs from here so we can keep track.

Good point! I have reactjs/react-codemod#122, reactjs/react-codemod#132, reactjs/react-codemod#156, and reactjs/react-codemod#157 open right now. Adding this to the original PR.

jsnmoon commented:
> One more bit of feedback: Since this PR contains a codemod to use the new standalone [`prop-types` npm](https://www.npmjs.com/package/prop-types), I think we should add it to `package.json`.

> Oh, same for [`create-react-class`](https://www.npmjs.com/package/create-react-class).

Done! `create-react-class` already existed as a dependency, so I went ahead and updated it to the latest version for good measure.

Rebased and re-wrapped. I'll aim to land this sometime this/next week!

ockham commented:
Re-based and re-shrunk one last time. As discussed with @jsnmoon in Slack, I'm going to merge this on his behalf!

jsnmoon commented:
Thanks so much @ockham! :)
